community/kernel/src/main/java/org/neo4j/kernel/impl/store/StoreVersionTrailerUtil.java

```diff
@@ -0,0 +1,132 @@
+            throws IOException
+    {
+        int trailerOffset;
+        byte[] encodedTrailer = UTF8.encode( trailer );
+        int maximumNumberOfPagesVersionSpans = getMaximumNumberOfPagesVersionSpans( encodedTrailer.length, pagedFile );
+        long pageId = Math.max( pagedFile.getLastPageId() + 1 - maximumNumberOfPagesVersionSpans, 0 );
+        try ( PageCursor pageCursor = pagedFile.io( pageId, PagedFile.PF_SHARED_LOCK ) )
+        {
+            byte[] allData = getLastBytes( pagedFile, maximumNumberOfPagesVersionSpans, pageCursor );
+            trailerOffset = findTrailerOffset( allData, encodedTrailer );
+        }
+        return trailerOffset + pageId * pagedFile.pageSize();
+    }
+
+    public static void writeTrailer( PagedFile pagedFile, byte[] trailer,
+            long trailerOffset ) throws IOException
+    {
+        long pageId = trailerOffset / pagedFile.pageSize();
+        try ( PageCursor pageCursor = pagedFile.io( pageId, PagedFile.PF_EXCLUSIVE_LOCK ) )
+        {
+            int writtenOffset = 0;
+            while ( writtenOffset < trailer.length )
+            {
+                pageCursor.next();
+                pageCursor.setOffset( (int) ((writtenOffset + trailerOffset) % pagedFile.pageSize()) );
+                do
+                {
+                    do
+                    {
+                        pageCursor.putByte( trailer[writtenOffset] );
+                    }
+                    while ( pageCursor.shouldRetry() );
+                    writtenOffset++;
+                }
+                while ( (writtenOffset + trailerOffset) % pagedFile.pageSize() != 0 &&
+                        writtenOffset < trailer.length );
+            }
+        }
+    }
+
+    public static String readVersion( PagedFile pagedFile, String expectedVersion ) throws IOException
+    {
+        int encodedExpectedVersion = UTF8.encode( expectedVersion ).length;
+        String version = null;
+        int maximumNumberOfPagesVersionSpans = getMaximumNumberOfPagesVersionSpans( encodedExpectedVersion, pagedFile );
+        long pageId = Math.max( pagedFile.getLastPageId() + 1 - maximumNumberOfPagesVersionSpans, 0 );
+        try ( PageCursor pageCursor = pagedFile.io( pageId, PagedFile.PF_SHARED_LOCK ) )
+        {
+            byte[] allData = getLastBytes( pagedFile, maximumNumberOfPagesVersionSpans, pageCursor );
+            int offset = findTrailerOffset( allData, UTF8.encode( expectedVersion.split( " " )[0] ) );
+            if ( offset != -1 )
+            {
+                version = new String( allData, offset, encodedExpectedVersion, Charsets.UTF_8 );
+            }
+        }
+        return version;
+    }
+
+    public static int findTrailerOffset( byte[] src, byte[] sought )
+    {
+        for ( int i = src.length - sought.length; i >= 0; i-- )
+        {
+            int pos = 0;
+            while ( pos < sought.length && src[i + pos] == sought[pos] )
```

Comment:
it might have been faster to just check `0`s and the start comparing arrays, perhaps not super important